### ISSUE TITLE
uvertTabellen: Kommentar

### ISSUE THREAD

**Issue by traeker** [bclausen/uvert#6]
Schöne Übersicht, hier einige Fragen/Anregungen/Ideen/Überlegungen:

- Wo erfolgt die Pluralbildung von RoR? Die meisten Tabellen sind im Plural formuliert? sind die Attribute von der automatischen Pluralbildung betroffen?

- Primärschlüssel: Ich plädiere dafür, dass wir uns mit den Primärschlüsseln an WinSchool halten und aussagekräftige Bezeichner verwenden. D.h. in den Tabellen classes, teachers, electives, projectgroups, terms, subjects haben wir bereits eindeutige Bezeichner (Lehrerkürzel, Fachkürzel, Kursbezeichnung, Name des Zeitraums usw.) Die Relationstabellen werden übersichtlicher, wenn wir mit diesen aussagekräftigen Primärschlüsseln arbeiten, dort steht dann bspw. RKR MAT 6d anstatt 37 5 8. Bei den Relationtabellen ist evtl. auch kein primärschlüssel nötig? Bin mir unsicher, die SQL-Workbench hat dort keinen erzeugt und mir fällt spontan auch keine Notwendigkeit ein. 

- In der Tabelle classes_classprofiles sind start- und Enddatum eingetragen. Besser wäre ein Bezug auf terms

- Wir müssen uns gut überlegen, wo/wie wir classprofiles aufhängen. So hängt es ein bisschen in der Luft. Wollen wir die Namen und die Stündigkeit der Fächer darin festhalten? Dann bräuchten wir sie nicht in classdistribution (ohne Plural-s (!)).

-in class_distribution benötigen dringend einen Bezug zu term_id. Es ist überhaupt die Frage, ob wir diese nicht in jeder (!?) Tabelle benötigen - und eher nciht in den reinen Relationen. Mmmh....nee, in teachers brauchen wir die nicht so explizit... dort könnte man eine Spalte start_term einrichten - mit Bezug auf terms

Insgesamt finde ich, dass in dem Datenbankschema sehr gute Ideen dargestellt sind. Über die o.g. Dinge müssen wir uns aber noch Gedanken machen.

**Comment by bclausen** [comment]
Die Pluralbildung bezieht sich nicht direkt auf die Tabellen sondern auf die Klassen (teacher, class, relief usw.). Da schreibt man dann "teacher has and belongs to many reliefs" (bei classes kann ich nicht so genau sagen, ob das funktioniert).

Was genau in den einzelnen Tabellen eingetragen wird, ist ja für die Views uninteressant. Wenn wir Lehrerkürzel, Fachkürzel usw. als Primärschlüssel verwenden, können diese nicht mehr geändert werden. Es ist besser wenn für Mathematik eine Zahl verwendet wird. So muss das Fach nur an einer Stelle geändert und nicht in allen Tabellen geändert werden. Die Tabellen müssen ja auch nicht für uns lesbar ausgefüllt werden. Darum kümmert sich RoR automatisch. Wenn man sich die Datenbank später ansieht, findet man nur IDs als Einträge.
Für das Schema, wäre es aber wohl übersichtlicher, mit lesbaren Einträgen zu arbeiten. Werde ich noch ändern.

Mit dem Start- und Enddatum in classes_classprofiles möchte ich festlegen wie lange das Profil für eine bestimmte Klasse gültig ist. Darüber regeln wir dann die Situation, wenn ein Lehrer nur eine bestimmte Zeit im Halbjahr eine Klasse unterrichtet. D.h. das wird im Profil festgelegt. So war meine Idee.
In terms will ich nur Anfang und Ende des Schulhalbjahres festlegen.

In der Tabell classprofiles können wir die Fächer nicht festhalten, da wir nicht wissen, wie viele Fächer ein Profil enthält. Es handelt sich ja um eine n:m Assoziation. dafür wird immer eine Hilfstabelle benötigt. Aber, ob es so funktioniert weiß ich auch noch nicht. Es wird nicht gelingen, das Datenbankschema schon exakt festzulegen. Es wird sich noch ändern während der weiteren Entwicklung.

Der Bezug von class_distribution zu term erfolgt über classes_classprofiles. Hier ist also ein Fehler in der Tabelle (habe es geändert). 

In der Datenbank soll ja Redundanz vermieden werden. daher ist nicht immer ein term-eintrag notwendig.

**Comment by traeker** [comment]
Ja, für die Views sind die Primärschlüssel uninteressant. Aber warum sollen wir neue Bezeichner einführen, wenn wir schon eindeutige Bezeichner haben? Wir können doch auch Werte von Primärschlüsseln ändern, solange sie eindeutig bleiben. Die Lehrerkürzel, Klassenbezeichnungen usw. sind doch natürliche und sinnstiftende Primärschlüssel.
